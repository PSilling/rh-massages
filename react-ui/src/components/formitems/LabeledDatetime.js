// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Col, FormGroup, Label, Tooltip } from "reactstrap";
import Datetime from "react-datetime";
import moment from "moment";

// util imports
import Util from "../../util/Util";

/**
 * A combined formgroup component with a label (can have a tooltip) and a Datetime input.
 * A column is included to set input size.
 */
class LabeledDatetime extends Component {
  state = { tooltipActive: false };

  tooltipTarget = Util.getTooltipTargets(1)[0];

  toggleTooltip = () => {
    this.setState(prevState => ({ tooltipActive: !prevState.tooltipActive }));
  };

  handleKeyPress = event => {
    if (event.key === "Enter") {
      this.props.onEnterPress();
    }
  };

  render() {
    const { disabled, onChange, label, value, labelWidth, onEnterPress, size, tooltip, ...rest } = this.props;
    return (
      <Col md={this.props.size}>
        <FormGroup>
          <Label id={this.tooltipTarget} for={`${this.tooltipTarget}_input`} style={{ width: this.props.labelWidth }}>
            {this.props.label}
          </Label>
          <Datetime
            {...rest}
            value={this.props.value}
            onChange={this.props.onChange}
            inputProps={{
              id: `${this.tooltipTarget}_input`,
              disabled,
              placeholder: this.props.label,
              onKeyPress: this.handleKeyPress
            }}
          />
          {this.props.tooltip !== "" && (
            <Tooltip isOpen={this.state.tooltipActive} target={this.tooltipTarget} toggle={this.toggleTooltip}>
              {this.props.tooltip}
            </Tooltip>
          )}
        </FormGroup>
      </Col>
    );
  }
}

LabeledDatetime.propTypes = {
  /** function called on input value change */
  onChange: PropTypes.func.isRequired,
  /** input label text */
  label: PropTypes.node.isRequired,
  /** current value of the input field */
  value: PropTypes.oneOfType([PropTypes.instanceOf(moment), PropTypes.instanceOf(Date)]).isRequired,
  /** whether the Datetime should be disabled */
  disabled: PropTypes.bool,
  /** width of the label container */
  labelWidth: PropTypes.string,
  /** input datalist options */
  onEnterPress: PropTypes.func,
  /** input column length */
  size: PropTypes.string,
  /** input label tooltip */
  tooltip: PropTypes.string
};

LabeledDatetime.defaultProps = {
  onEnterPress() {},
  disabled: false,
  labelWidth: "auto",
  size: "12",
  tooltip: ""
};

export default LabeledDatetime;
