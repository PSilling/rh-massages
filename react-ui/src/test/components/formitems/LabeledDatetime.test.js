// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import { Col, Label, Tooltip } from "reactstrap";
import Datetime from "react-datetime";
import moment from "moment";
import LabeledDatetime from "../../../components/formitems/LabeledDatetime";

test("renders content with correct props", () => {
  const testFunction = jest.fn();
  const wrapper = shallow(<LabeledDatetime onChange={testFunction} label="test" value={moment(1)} tooltip="tooltip" />);
  const col = wrapper.find(Col);
  const label = wrapper.find(Label);
  const datetime = wrapper.find(Datetime);
  const tooltip = wrapper.find(Tooltip);

  expect(testFunction).not.toHaveBeenCalled();

  datetime.props().onChange();

  expect(testFunction).toHaveBeenCalledTimes(1);

  expect(col.props().md).toEqual("12");
  expect(label.props().id).toEqual(tooltip.props().target);
  expect(label.props().children).toEqual("test");
  expect(datetime.props().value).toEqual(moment(1));
  expect(tooltip.props().isOpen).toBe(false);
  expect(tooltip.props().children).toEqual("tooltip");
  expect(wrapper).toMatchSnapshot();
});
