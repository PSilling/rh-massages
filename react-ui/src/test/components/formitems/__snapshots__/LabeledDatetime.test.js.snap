// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content with correct props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LabeledDatetime
    disabled={false}
    label="test"
    onChange={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEnterPress={[Function]}
    size="12"
    tooltip="tooltip"
    value={"1970-01-01T00:00:00.001Z"}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <FormGroup
        tag="div"
      >
        <Label
          for="Tooltip1_input"
          id="Tooltip1"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          test
        </Label>
        <DateTime
          className=""
          closeOnSelect={false}
          closeOnTab={true}
          dateFormat={true}
          defaultValue=""
          input={true}
          inputProps={
            Object {
              "disabled": false,
              "id": "Tooltip1_input",
              "onKeyPress": [Function],
              "placeholder": "test",
            }
          }
          onBlur={[Function]}
          onChange={
            [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            }
          }
          onFocus={[Function]}
          onNavigateBack={[Function]}
          onNavigateForward={[Function]}
          onViewModeChange={[Function]}
          strictParsing={true}
          timeConstraints={Object {}}
          timeFormat={true}
          utc={false}
          value={"1970-01-01T00:00:00.001Z"}
        />
        <Tooltip
          autohide={true}
          delay={
            Object {
              "hide": 250,
              "show": 0,
            }
          }
          hideArrow={false}
          isOpen={false}
          placement="top"
          placementPrefix="bs-tooltip"
          target="Tooltip1"
          toggle={[Function]}
        >
          tooltip
        </Tooltip>
      </FormGroup>,
      "md": "12",
      "tag": "div",
      "widths": Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Label
            for="Tooltip1_input"
            id="Tooltip1"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            test
          </Label>,
          <DateTime
            className=""
            closeOnSelect={false}
            closeOnTab={true}
            dateFormat={true}
            defaultValue=""
            input={true}
            inputProps={
              Object {
                "disabled": false,
                "id": "Tooltip1_input",
                "onKeyPress": [Function],
                "placeholder": "test",
              }
            }
            onBlur={[Function]}
            onChange={
              [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            onFocus={[Function]}
            onNavigateBack={[Function]}
            onNavigateForward={[Function]}
            onViewModeChange={[Function]}
            strictParsing={true}
            timeConstraints={Object {}}
            timeFormat={true}
            utc={false}
            value={"1970-01-01T00:00:00.001Z"}
          />,
          <Tooltip
            autohide={true}
            delay={
              Object {
                "hide": 250,
                "show": 0,
              }
            }
            hideArrow={false}
            isOpen={false}
            placement="top"
            placementPrefix="bs-tooltip"
            target="Tooltip1"
            toggle={[Function]}
          >
            tooltip
          </Tooltip>,
        ],
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "test",
            "for": "Tooltip1_input",
            "id": "Tooltip1",
            "tag": "label",
            "widths": Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ],
          },
          "ref": null,
          "rendered": "test",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "",
            "closeOnSelect": false,
            "closeOnTab": true,
            "dateFormat": true,
            "defaultValue": "",
            "input": true,
            "inputProps": Object {
              "disabled": false,
              "id": "Tooltip1_input",
              "onKeyPress": [Function],
              "placeholder": "test",
            },
            "onBlur": [Function],
            "onChange": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            },
            "onFocus": [Function],
            "onNavigateBack": [Function],
            "onNavigateForward": [Function],
            "onViewModeChange": [Function],
            "strictParsing": true,
            "timeConstraints": Object {},
            "timeFormat": true,
            "utc": false,
            "value": "1970-01-01T00:00:00.001Z",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autohide": true,
            "children": "tooltip",
            "delay": Object {
              "hide": 250,
              "show": 0,
            },
            "hideArrow": false,
            "isOpen": false,
            "placement": "top",
            "placementPrefix": "bs-tooltip",
            "target": "Tooltip1",
            "toggle": [Function],
          },
          "ref": null,
          "rendered": "tooltip",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <FormGroup
          tag="div"
        >
          <Label
            for="Tooltip1_input"
            id="Tooltip1"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            test
          </Label>
          <DateTime
            className=""
            closeOnSelect={false}
            closeOnTab={true}
            dateFormat={true}
            defaultValue=""
            input={true}
            inputProps={
              Object {
                "disabled": false,
                "id": "Tooltip1_input",
                "onKeyPress": [Function],
                "placeholder": "test",
              }
            }
            onBlur={[Function]}
            onChange={
              [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            onFocus={[Function]}
            onNavigateBack={[Function]}
            onNavigateForward={[Function]}
            onViewModeChange={[Function]}
            strictParsing={true}
            timeConstraints={Object {}}
            timeFormat={true}
            utc={false}
            value={"1970-01-01T00:00:00.001Z"}
          />
          <Tooltip
            autohide={true}
            delay={
              Object {
                "hide": 250,
                "show": 0,
              }
            }
            hideArrow={false}
            isOpen={false}
            placement="top"
            placementPrefix="bs-tooltip"
            target="Tooltip1"
            toggle={[Function]}
          >
            tooltip
          </Tooltip>
        </FormGroup>,
        "md": "12",
        "tag": "div",
        "widths": Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Label
              for="Tooltip1_input"
              id="Tooltip1"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              test
            </Label>,
            <DateTime
              className=""
              closeOnSelect={false}
              closeOnTab={true}
              dateFormat={true}
              defaultValue=""
              input={true}
              inputProps={
                Object {
                  "disabled": false,
                  "id": "Tooltip1_input",
                  "onKeyPress": [Function],
                  "placeholder": "test",
                }
              }
              onBlur={[Function]}
              onChange={
                [MockFunction] {
                  "calls": Array [
                    Array [],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
              onFocus={[Function]}
              onNavigateBack={[Function]}
              onNavigateForward={[Function]}
              onViewModeChange={[Function]}
              strictParsing={true}
              timeConstraints={Object {}}
              timeFormat={true}
              utc={false}
              value={"1970-01-01T00:00:00.001Z"}
            />,
            <Tooltip
              autohide={true}
              delay={
                Object {
                  "hide": 250,
                  "show": 0,
                }
              }
              hideArrow={false}
              isOpen={false}
              placement="top"
              placementPrefix="bs-tooltip"
              target="Tooltip1"
              toggle={[Function]}
            >
              tooltip
            </Tooltip>,
          ],
          "tag": "div",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "test",
              "for": "Tooltip1_input",
              "id": "Tooltip1",
              "tag": "label",
              "widths": Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ],
            },
            "ref": null,
            "rendered": "test",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "",
              "closeOnSelect": false,
              "closeOnTab": true,
              "dateFormat": true,
              "defaultValue": "",
              "input": true,
              "inputProps": Object {
                "disabled": false,
                "id": "Tooltip1_input",
                "onKeyPress": [Function],
                "placeholder": "test",
              },
              "onBlur": [Function],
              "onChange": [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              },
              "onFocus": [Function],
              "onNavigateBack": [Function],
              "onNavigateForward": [Function],
              "onViewModeChange": [Function],
              "strictParsing": true,
              "timeConstraints": Object {},
              "timeFormat": true,
              "utc": false,
              "value": "1970-01-01T00:00:00.001Z",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autohide": true,
              "children": "tooltip",
              "delay": Object {
                "hide": 250,
                "show": 0,
              },
              "hideArrow": false,
              "isOpen": false,
              "placement": "top",
              "placementPrefix": "bs-tooltip",
              "target": "Tooltip1",
              "toggle": [Function],
            },
            "ref": null,
            "rendered": "tooltip",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
